malariagen_data/ag3.py

```diff
@@ -2489,6 +2490,92 @@ def aa_allele_frequencies(
 
+    def plot_frequencies_heatmap(
+        self, df, index, index_name, max_len=100, width=500, colorbar=True
+    ):
+
+        """Generate a Plotly.express frequency heatmap using pandas DataFrame output from ag3.snp_allele_frequencies()
+        or ag3.gene_cnv_frequencies() that has been user-filtered to just rows of interest (fewer rows than max_len).
+
+        Parameters
+        ----------
+        df : pandas DataFrame
+           Filtered output from ag3.snp_allele_frequencies() or ag3.gene_cnv_frequencies.
+        index : list of str
+            A list of one or more column headers present in the input dataframe, these become
+            the heatmap y-axis row labels. When taken together, these columns must produce a unique index.
+        index_name : str
+            This is the y-axis label that will be displayed on the heatmap
+        max_len : int, optional
+            Displaying large styled dataframes may cause ipython notebooks to crash.
+        width : int
+            Width of styled output.
+        colorbar : bool, optional
+            If False, colorbar is not output.
+
+        Returns
+        -------
+        Plotly.express heatmap built from the index and "frq_" columns of input DataFrame, where
+        increasing frequency is denoted by intensity of red coloration.
+
+        """
+        # check len of input
+        assert len(df) <= max_len
+
+        # reset index
+        df = df.reset_index().copy()
+
+        # ensure that the things we want to make the index out of are all str type
+        for i in index:
+            df.loc[:, i] = df[i].astype(str)
+
+        # make a new column out of index list and give it the index_name
+        df.loc[:, index_name] = df[index].agg("_".join, axis=1)
+
+        # check that index is unique (otherwise style won't work)
+        assert df[index_name].nunique() == len(
+            df
+        ), f"{index} - does not produce a unique index"
+
+        # drop and re-order columns
+        frq_cols = [index_name] + sorted([col for col in df.columns if "frq_" in col])
+
+        # check that freqs are all floats
+        for i in [col for col in df.columns if "frq_" in col]:
+            df.loc[:, i] = df[i].astype(float)
+
+        # keep only freq cols
+        heatmap_df = df[frq_cols].copy()
+
+        # set index
+        heatmap_df.set_index(index_name, inplace=True)
+
+        # clean column names
+        heatmap_df.columns = heatmap_df.columns.str.lstrip("frq_")
+
+        # plotly heatmap styling
+        fig = px.imshow(
+            img=df,
+            zmin=0,
+            zmax=1,
+            text_auto=".0%",
+            aspect="auto",
+            width=width,
+            color_continuous_scale="Reds",
+        )
+
+        fig.update_xaxes(side="top", tickangle=270, title="cohorts")
+        fig.update_layout(
+            coloraxis_colorbar=dict(
+                title="frequency",
+                tickvals=[0, 0.2, 0.4, 0.6, 0.8, 0.99],
```

Comment:
Why doesn't this end with 1.0?

Reply:
Aesthetics - I was trying to imperceptibly move the final tickval away from the colorbar title.

Reply:
Completely, unnecessary of course!